### PR TITLE
Search 161

### PR DESCRIPTION
**Pull request by alexbakharew**
Draft PR

### CODE REVIEW

File: core/search/levenshtein_filter.hpp [hunk 1]
```diff
@@ -61,16 +61,21 @@ struct IRESEARCH_API by_edit_distance_filter_options {
   //////////////////////////////////////////////////////////////////////////////
   bool with_transpositions{false};
 
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief compute levenshtein without prefix
```
Comment (gnusi) [inline core/search/levenshtein_filter.hpp:65]: match this number of characters from the beginning of the target regardless of edit distance

File: core/utils/levenshtein_utils.cpp [hunk 2]
```diff
@@ -613,18 +614,31 @@ automaton make_levenshtein_automaton(
...
 
   // initial state
-  a.SetStart(a.AddState());
+  auto start_state = a.AddState();
+  a.SetStart(start_state);
+
+  auto begin = prefix.begin();
+  auto end = prefix.end();
+  decltype(start_state) to;
+  for (; begin != prefix.end(); ) {
```
Comment (gnusi) [inline core/utils/levenshtein_utils.cpp:623]:
```suggestion
for (; begin != end; ) {
```

File: tests/index/index_levenshtein_tests.cpp [hunk 3]
```diff
@@ -101,11 +102,12 @@ TEST_P(levenshtein_automaton_index_test_case, test_lev_automaton) {
   auto reader = open_reader();
   ASSERT_NE(nullptr, reader);
 
+  irs::bstring prefix; // empty prefix
   for (auto& description : DESCRIPTIONS) {
     for (auto& target : TARGETS) {
       SCOPED_TRACE(testing::Message("Target: '") << target <<
                    testing::Message("', Edit distance: ") << size_t(description.max_distance()));
-      assert_index(reader, description, irs::ref_cast<irs::byte_type>(target));
+      assert_index(reader, description,  irs::ref_cast<irs::byte_type>(prefix), irs::ref_cast<irs::byte_type>(target));
```
Comment (gnusi) [inline tests/index/index_levenshtein_tests.cpp:110]: we can use `bytes_ref::EMPTY` instead

File: core/utils/levenshtein_utils.cpp [hunk 4]
```diff
@@ -613,18 +614,31 @@ automaton make_levenshtein_automaton(
...
+    const byte_type* next = irs::utf8_utils::next(begin, end);
```
Comment (gnusi) [inline core/utils/levenshtein_utils.cpp:624]: we don't need to use `irs::` namespace prefix in `irs` namespace

File: tests/utils/levenshtein_utils_test.cpp [hunk 5]
```diff
@@ -90,6 +94,14 @@ void assert_description(
   }
 }
 
+void assert_description(
+    const irs::parametric_description& description,
+    const irs::bytes_ref& target,
+    const std::vector<std::tuple<irs::bytes_ref, size_t, size_t, size_t>>& candidates) {
+
+  return assert_description(description, irs::ref_cast<irs::byte_type>(""_sr), target, candidates);
```
Comment (gnusi) [inline tests/utils/levenshtein_utils_test.cpp:102]: `irs::ref_cast<irs::byte_type>(""_sr)` -> `bytes_ref::EMPTY`

File: core/search/levenshtein_filter.hpp [hunk 6]
```diff
@@ -61,16 +61,21 @@ struct IRESEARCH_API by_edit_distance_filter_options {
   //////////////////////////////////////////////////////////////////////////////
   bool with_transpositions{false};
 
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief match this number of characters from the beginning of the target regardless of edit distance
```
Comment (gnusi) [inline core/search/levenshtein_filter.hpp:65]: 80 characters per line

File: core/search/levenshtein_filter.hpp [hunk 7]
```diff
@@ -61,16 +61,21 @@ struct IRESEARCH_API by_edit_distance_filter_options {
...
+  //////////////////////////////////////////////////////////////////////////////
+  bstring prefix;
```
Comment (gnusi) [inline core/search/levenshtein_filter.hpp:67]: the good practice is to sort structure member in descending order of their size, i.e. you have to put `prefix` right after `target`

File: core/search/levenshtein_filter.cpp [hunk 8]
```diff
@@ -267,23 +271,28 @@ DEFINE_FACTORY_DEFAULT(by_edit_distance)
...
-    opts.max_distance, opts.provider, opts.with_transpositions,
+    opts.max_distance, opts.provider, opts.with_transpositions, opts.prefix, opts.term, 
     []() -> field_visitor {
       return [](const sub_reader&, const term_reader&, filter_visitor&){};
     },
     [&opts]() -> field_visitor {
+      auto target = opts.prefix + opts.term;
       // must copy term as it may point to temporary string
-      return [term = opts.term](
+      return [target](
```
Comment (gnusi) [inline core/search/levenshtein_filter.cpp:281]:
```suggestion
return [target = opts.prefix + opts.term](
```

File: core/search/levenshtein_filter.cpp [hunk 9]
```diff
@@ -321,18 +330,24 @@ DEFINE_FACTORY_DEFAULT(by_edit_distance)
...
-    max_distance, provider, with_transpositions,
+    max_distance, provider, with_transpositions, prefix, term,
     []() -> filter::prepared::ptr {
       return prepared::empty();
     },
-    [&index, &order, boost, &field, &term]() -> filter::prepared::ptr {
-      return by_term::prepare(index, order, boost, field, term);
+    [&index, &order, boost, &field, &prefix, &term]() -> filter::prepared::ptr {
+      bstring target(prefix.c_str(), prefix.size());
+      target += term;
```
Comment (gnusi) [inline core/search/levenshtein_filter.cpp:343]:
```
if (!prefix.empty() && !term.empty()) {
  bstring target;
  target.reserve(prefix.size() + term.size();
  target += prefix;
  target += term;
  return by_term::prepare(index, order, boost, field, target);
} 

return by_term::prepare(index, order, boost, field, prefix.empty() ? term : prefix);
```

**Review by gnusi — CHANGES_REQUESTED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM

**Review by gnusi — APPROVED** [review]

**Pull request merged**